akka-http-core/src/main/scala/akka/http/scaladsl/model/HttpMessage.scala

```diff
@@ -106,13 +109,30 @@ sealed trait HttpMessage extends jm.HttpMessage {
   /** Returns the first header of the given type if there is one */
   def header[T >: Null <: jm.HttpHeader: ClassTag]: Option[T] = {
     val clazz = classTag[T].runtimeClass.asInstanceOf[Class[T]]
-    HttpHeader.fastFind[T](clazz, headers) match {
+    if (classOf[akka.http.scaladsl.model.headers.ModeledCustomHeader[T]].isAssignableFrom(clazz)) {
+      findHeaderViaModeledCompanion(clazz).asInstanceOf[Option[T]]
+    } else HttpHeader.fastFind[T](clazz, headers) match {
       case OptionVal.Some(h)                     ⇒ Some(h)
       case _ if clazz == classOf[`Content-Type`] ⇒ Some(`Content-Type`(entity.contentType)).asInstanceOf[Option[T]]
       case _                                     ⇒ None
     }
   }
 
+  private def findHeaderViaModeledCompanion(clazz: Class[_]) = {
+    // allows locating "companion" without implicits, same trick is applied in javadsl
+    val refl = new ReflectiveDynamicAccess(getClass.getClassLoader)
+    refl.getObjectFor[ModeledCustomHeaderCompanion[_]](clazz.getName) match {
+      case Success(companion) ⇒
+        headers.collectFirst {
+          case RawHeader(name, value) if name.toLowerCase(Locale.ROOT) == companion.lowercaseName ⇒ companion(value)
```

Comment:
Parsing here introduces a new failure condition for the method which I'd rather not like to introduce. Also it's inconsistent with what we do for regular modeled headers which won't be found if they only exist as a `RawHeader`.